src/hammer-vlsi/par/innovus/__init__.py

```diff
@@ -0,0 +1,249 @@
+
+        # GDS streamout.
+        verbose_append("write_stream -output_macros -mode ALL -unit 1000 gds_file")
+        # extra junk: -map_file -attach_inst_name ... -attach_net_name ...
+
+        # Quit Innovus.
+        verbose_append("exit")
+
+        # Create par script.
+        par_tcl_filename = os.path.join(self.run_dir, "par.tcl")
+        with open(par_tcl_filename, "w") as f:
+            f.write("\n".join(output))
+
+        # Make sure that generated-scripts exists.
+        generated_scripts_dir = os.path.join(self.run_dir, "generated-scripts")
+        os.makedirs(generated_scripts_dir, exist_ok=True)
+
+        # Create open_chip script.
+        with open(os.path.join(generated_scripts_dir, "open_chip.tcl"), "w") as f:
+            f.write("""
+read_db {name}
+        """.format(name=output_innovus_lib_name))
+
+        with open(os.path.join(generated_scripts_dir, "open_chip"), "w") as f:
+            f.write("""
+cd {run_dir}
+source enter
+$INNOVUS_BIN -common_ui -files generated-scripts/open_chip.tcl
+        """.format(run_dir=self.run_dir))
+        self.run_executable([
+            "chmod", "+x", os.path.join(generated_scripts_dir, "open_chip")
+        ])
+
+        # Build args.
+        args = [
+            self.get_setting("par.innovus.innovus_bin"),
+            "-nowin",  # Prevent the GUI popping up.
+            "-common_ui",
+            "-files", par_tcl_filename
+        ]
+
+        # Temporarily disable colours/tag to make run output more readable.
+        # TODO: think of a more elegant way to do this?
+        HammerVLSILogging.enable_colour = False
+        HammerVLSILogging.enable_tag = False
+        self.run_executable(args, cwd=self.run_dir)  # TODO: check for errors and deal with them
+        HammerVLSILogging.enable_colour = True
+        HammerVLSILogging.enable_tag = True
+
+        # TODO: check that par run was successful
+
+        return True
+
+    def create_floorplan_tcl(self) -> List[str]:
+        """
+        Create a floorplan TCL depending on the floorplan mode.
+        """
+        output = []  # type: List[str]
+
+        floorplan_mode = str(self.get_setting("par.innovus.floorplan_mode"))
+        if floorplan_mode == "manual":
+            floorplan_script_contents = str(self.get_setting("par.innovus.floorplan_script_contents"))
+            # TODO(edwardw): proper source locators/SourceInfo
+            output.append("# Floorplan manually specified from HAMMER")
+            output.extend(floorplan_script_contents.split("\n"))
+        elif floorplan_mode == "generate":
+            output.extend(self.generate_floorplan_tcl())
+        else:
+            if floorplan_mode != "blank":
+                self.logger.error("Invalid floorplan_mode {mode}. Using blank floorplan.".format(mode=floorplan_mode))
+            # Write blank floorplan
+            output.append("# Blank floorplan specified from HAMMER")
+        return output
+
+    def generate_mmmc_script(self) -> str:
```

Comment:
oh I see my earlier comment was silly given this exists. But this isn't really an innovus thing but rather a place and route tool thing

Reply:
(see mmmc comment)